fsm.go

```diff
@@ -28,20 +28,38 @@ const (
 	maximumRetries             = 3
 )
 
+type fsmAgent interface {
+	getHost() string
+	setHost(host string)
+	makeURL(prefix string) string
+	makeHostURL(host string, prefix string) string
+	applyHostAgentSettings(resp agentResponse)
+	requestHeader(url string, method string, header string) (string, error)
+	announceRequest(url string, method string, data interface{}, ret *agentResponse) (string, error)
+	head(url string) (string, error)
+}
+
 type fsmS struct {
-	agent       *agentS
-	fsm         *f.FSM
-	timer       *time.Timer
-	retriesLeft int
+	name                       string
+	agent                      fsmAgent
+	fsm                        *f.FSM
+	timer                      *time.Timer
+	retriesLeft                int
+	expDelayFunc               func(retryNumber int) time.Duration
+	lookupAgentHostRetryPeriod time.Duration
+	logger                     LeveledLogger
 }
 
-func newFSM(agent *agentS) *fsmS {
-	agent.logger.Warn("Stan is on the scene. Starting Instana instrumentation.")
-	agent.logger.Debug("initializing fsm")
+func newFSM(agent fsmAgent, logger LeveledLogger) *fsmS {
+	logger.Warn("Stan is on the scene. Starting Instana instrumentation.")
+	logger.Debug("initializing fsm")
 
 	ret := &fsmS{
-		agent:       agent,
-		retriesLeft: maximumRetries,
+		agent:                      agent,
+		retriesLeft:                maximumRetries,
+		expDelayFunc:               expDelay,
```

Comment:
Query : Here I was looking from where `expDelayFunc` was taking value. Does it make sense to move the `expDelay` to the top of the file. Or is it preferred to the keep the constructor function at the top?

Reply:
`expDelay` is used in multiple places, it doesn't belong to this particular type. It is also used in other places.